Canadian build for bare-metal ARM fails.

Using crosstool-ng from github master (after 1.20)

I want to build a version of GCC to run on the Raspberry Pi, and target the Teensy 3.1 (a Cortex-M4 without FPU.) My build machine is Linux amd64.

1) Configure and build cross from x64 to rpi -- works
2) Configure and build cross from x64 to teensy (bare metal, newlib, no long double or quadmath) -- works, after patching the "can't determine value of EOF" problem
3) Configure and build canadian from rpi to teensy -- compiler builds, but library fails with file not found from an #include_next <fenv.h>

Internet search says this is a "cygwin specific problem that requires a newer cygwin," but that's obviously not the problem in this case.

    [ALL  ]    mkdir -p ./arm-jwatte-eabi/bits/stdc++.h.gch
    [ALL  ]    arm-jwatte-eabi-c++     -x c++-header -nostdinc++ -g -Os -I/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include/arm-jwatte-eabi -I/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include -I/usr/local/src/cross-source-target/rpi-teensy/.build/src/gcc-linaro-4.8-2014.04/libstdc++-v3/libsupc++ -I/opt/cross/rpi/include -O2 -g -std=gnu++0x /usr/local/src/cross-source-target/rpi-teensy/.build/src/gcc-linaro-4.8-2014.04/libstdc++-v3/include/precompiled/stdc++.h   -o arm-jwatte-eabi/bits/stdc++.h.gch/O2ggnu++0x.gch
    [ALL  ]    arm-jwatte-eabi-c++     -x c++-header -nostdinc++ -g -Os -I/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include/arm-jwatte-eabi -I/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include -I/usr/local/src/cross-source-target/rpi-teensy/.build/src/gcc-linaro-4.8-2014.04/libstdc++-v3/libsupc++ -I/opt/cross/rpi/include -O2 -g /usr/local/src/cross-source-target/rpi-teensy/.build/src/gcc-linaro-4.8-2014.04/libstdc++-v3/include/precompiled/stdc++.h -o arm-jwatte-eabi/bits/stdc++.h.gch/O2g.gch
    [ALL  ]    In file included from /usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include/cfenv:41:0,
    [ALL  ]                     from /usr/local/src/cross-source-target/rpi-teensy/.build/src/gcc-linaro-4.8-2014.04/libstdc++-v3/include/precompiled/stdc++.h:53:
    [ERROR]    /usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include/fenv.h:36:24: fatal error: fenv.h: No such file or directory
    [ALL  ]     # include_next <fenv.h>
    [ALL  ]                            ^
    [ALL  ]    compilation terminated.
    [ERROR]    make[4]: *** [arm-jwatte-eabi/bits/stdc++.h.gch/O2ggnu++0x.gch] Error 1
    [ALL  ]    make[4]: *** Waiting for unfinished jobs....
    [ALL  ]    make[4]: Leaving directory `/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include'
    [ERROR]    make[3]: *** [all-recursive] Error 1
    [ALL  ]    make[3]: Leaving directory `/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3'
    [ERROR]    make[2]: *** [all] Error 2
    [ALL  ]    make[2]: Leaving directory `/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3'
    [ERROR]    make[1]: *** [all-target-libstdc++-v3] Error 2
    [ALL  ]    make[1]: Leaving directory `/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final'
    [ERROR]
    [ERROR]  >>
    [ERROR]  >>  Build failed in step 'Installing final compiler'
    [ERROR]  >>        called in step '(top-level)'
    [ERROR]  >>
    [ERROR]  >>  Error happened in: CT_DoExecLog[scripts/functions@257]
    [ERROR]  >>        called from: do_cc_core_backend[scripts/build/cc/gcc.sh@486]
    [ERROR]  >>        called from: do_cc_for_host[scripts/build/cc/gcc.sh@598]
    [ERROR]  >>        called from: main[scripts/crosstool-NG.sh@638]
    [ERROR]  >>
    [ERROR]  >>  For more info on this error, look at the file: 'build.log'
    [ERROR]  >>  There is a list of known issues, some with workarounds, in:
    [ERROR]  >>      '/usr/local/share/doc/crosstool-ng/ct-ng.1.20.0/B - Known issues.txt'
    [ERROR]
    [ERROR]  (elapsed: 42:22.64)

Nope, same problem with those patches applied :-(
Note that the same target build works when building cross, rather than Canadian. Don't know what that's about.
Also, it's trying to #include another fenv.h (presumably from the system,) which doesn't actually exist, as it's bare-metal.


    [ALL  ]    make[3]: Entering directory `/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3'
    [ALL  ]    Making all in include
    [ALL  ]    make[4]: Entering directory `/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include'
    [ALL  ]    mkdir -p ./arm-jwatte-eabi/bits/stdc++.h.gch
    [ALL  ]    mkdir -p ./arm-jwatte-eabi/bits/stdc++.h.gch
    [ALL  ]    arm-jwatte-eabi-c++     -x c++-header -nostdinc++ -g -Os -I/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include/arm-jwatte-eabi -I/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include -I/usr/local/src/cross-source-target/rpi-teensy/.build/src/gcc-linaro-4.8-2014.04/libstdc++-v3/libsupc++ -I/opt/cross/rpi/include -O2 -g -std=gnu++0x /usr/local/src/cross-source-target/rpi-teensy/.build/src/gcc-linaro-4.8-2014.04/libstdc++-v3/include/precompiled/stdc++.h   -o arm-jwatte-eabi/bits/stdc++.h.gch/O2ggnu++0x.gch
    [ALL  ]    arm-jwatte-eabi-c++     -x c++-header -nostdinc++ -g -Os -I/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include/arm-jwatte-eabi -I/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include -I/usr/local/src/cross-source-target/rpi-teensy/.build/src/gcc-linaro-4.8-2014.04/libstdc++-v3/libsupc++ -I/opt/cross/rpi/include -O2 -g /usr/local/src/cross-source-target/rpi-teensy/.build/src/gcc-linaro-4.8-2014.04/libstdc++-v3/include/precompiled/stdc++.h -o arm-jwatte-eabi/bits/stdc++.h.gch/O2g.gch
    [ALL  ]    In file included from /usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include/cfenv:41:0,
    [ALL  ]                     from /usr/local/src/cross-source-target/rpi-teensy/.build/src/gcc-linaro-4.8-2014.04/libstdc++-v3/include/precompiled/stdc++.h:53:
    [ERROR]    /usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include/fenv.h:36:24: fatal error: fenv.h: No such file or directory
    [ALL  ]     # include_next <fenv.h>
    [ALL  ]                            ^
    [ALL  ]    compilation terminated.
    [ERROR]    make[4]: *** [arm-jwatte-eabi/bits/stdc++.h.gch/O2ggnu++0x.gch] Error 1
    [ALL  ]    make[4]: *** Waiting for unfinished jobs....
    [ALL  ]    make[4]: Leaving directory `/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3/include'
    [ERROR]    make[3]: *** [all-recursive] Error 1
    [ALL  ]    make[3]: Leaving directory `/usr/local/src/cross-source-target/rpi-teensy/.build/arm-jwatte-eabi/build/build-cc-final/arm-jwatte-eabi/libstdc++-v3'

--- a/scripts/build/cc/gcc.sh
+++ b/scripts/build/cc/gcc.sh
@@ -205,20 +205,16 @@ do_cc_core_backend() {
             extra_config+=("--enable-threads=no")
             extra_config+=("--disable-shared")
             extra_user_config=( "${CT_CC_CORE_EXTRA_CONFIG_ARRAY[@]}" )
-            copy_headers=y  # For baremetal, as there's no headers to copy,
-                            # we copy an empty directory. So, who cares?
             ;;
         shared)
             extra_config+=("--enable-shared")
             extra_user_config=( "${CT_CC_CORE_EXTRA_CONFIG_ARRAY[@]}" )
-            copy_headers=y
             ;;
         baremetal)
             extra_config+=("--with-newlib")
             extra_config+=("--enable-threads=no")
             extra_config+=("--disable-shared")
             extra_user_config=( "${CT_CC_EXTRA_CONFIG_ARRAY[@]}" )
-            copy_headers=n
             ;;
         *)
             CT_Abort "Internal Error: 'mode' must be one of: 'static', 'shared' or 'baremetal', not '${mode:-(empty)}'"
@@ -231,10 +227,8 @@ do_cc_core_backend() {
         [ -n "${CT_TOOLCHAIN_BUGURL}" ] && extra_config+=("--with-bugurl=${CT_TOOLCHAIN_BUGURL}")
     fi
 
-    if [ "${copy_headers}" = "y" ]; then
-        CT_DoLog DEBUG "Copying headers to install area of bootstrap gcc, so it can build libgcc2"
-        CT_DoExecLog ALL cp -a "${CT_HEADERS_DIR}" "${prefix}/${CT_TARGET}/include"
-    fi
+    CT_DoLog DEBUG "Copying headers to install area of core C compiler"
+    CT_DoExecLog ALL cp -a "${CT_HEADERS_DIR}" "${prefix}/${CT_TARGET}/include"
 
     for tmp in ARCH ABI CPU TUNE FPU FLOAT; do
         eval tmp="\${CT_ARCH_WITH_${tmp}}"

--- a/scripts/build/libc/newlib.sh
+++ b/scripts/build/libc/newlib.sh
@@ -50,16 +50,16 @@ do_libc_check_config() {
 }
 
 do_libc_start_files() {
+    CT_DoStep INFO "Installing C library headers & start files"
+    CT_DoExecLog ALL cp -a "${CT_SRC_DIR}/newlib-${CT_LIBC_VERSION}/newlib/libc/include/." \
+    "${CT_HEADERS_DIR}"
     if [ "${CT_ATMEL_AVR32_HEADERS}" = "y" ]; then
-        CT_DoStep INFO "Installing C library headers & start files"
-
         CT_DoLog EXTRA "Installing Atmel's AVR32 headers"
         CT_DoExecLog ALL mkdir -p "${CT_PREFIX_DIR}/${CT_TARGET}/include"
         CT_DoExecLog ALL cp -r "${CT_SRC_DIR}/${LIBC_NEWLIB_AVR_HDRS_BASE}/avr32"   \
                                "${CT_PREFIX_DIR}/${CT_TARGET}/include/"
-
-        CT_EndStep
     fi
+    CT_EndStep
 }
 
 do_libc() {

